Frameworks/CoreText/DWriteWrapper_CoreText.mm

```diff
@@ -522,4 +523,58 @@ HRESULT STDMETHODCALLTYPE GetPixelsPerDip(_In_opt_ void* clientDrawingContext, _
     }
 
     return frame;
+}
+
+static CGSize _DWriteGetSize(CFAttributedStringRef string, CFRange range, CGSize maxSize, CFRange* fitRange) {
+    CGSize ret = CGSizeZero;
+    if (range.length == 0L) {
+        range.length = CFAttributedStringGetLength(string);
+    }
+
+    if (!string || range.length <= 0L) {
+        return ret;
+    }
+
+    ComPtr<IDWriteTextLayout> textLayout;
+    if (FAILED(__DWriteTextLayoutCreate(string, range, { CGPointZero, maxSize }, &textLayout))) {
+        return ret;
+    }
+
+    DWRITE_TEXT_METRICS textMetrics;
+    if (FAILED(textLayout->GetMetrics(&textMetrics))) {
+        return ret;
+    }
+
+    // TODO:: find more precise value than 1.0 to increase width by to fully enclose frame
+    ret.width = std::min(maxSize.width, textMetrics.widthIncludingTrailingWhitespace + 1.0f);
+    ret.height = std::min(maxSize.height, textMetrics.height);
+
+    if (fitRange) {
+        *fitRange = { range.location, 0L };
+        uint32_t lineCount = 0;
+
+        // Should return E_NOT_SUFFICIENT_BUFFER and popluate lineCount
+        if (textLayout->GetLineMetrics(nullptr, 0, &lineCount) != E_NOT_SUFFICIENT_BUFFER) {
+            return ret;
+        }
+
+        std::vector<DWRITE_LINE_METRICS> metrics(lineCount);
+        if (FAILED(textLayout->GetLineMetrics(metrics.data(), lineCount, &lineCount))) {
+            return ret;
+        }
+
+        float totalHeight = 0;
+        CFIndex endPos = range.location;
+        for (size_t i = 0; i < lineCount; ++i) {
```

Comment:
extra nit: const auto&

---
In reply to: [93963412](https://github.com/Microsoft/WinObjC/pull/1603#discussion_r93963412) [](ancestors = 93963412)